winch/codegen/src/codegen/context.rs

```diff
@@ -218,6 +218,19 @@ impl<'a, 'builtins> CodeGenContext<'a, 'builtins> {
         self.stack.push(typed_reg.into());
     }
 
+    /// Prepares arguments for emitting an f32 or f64 binary operation.
+    pub fn float_binop<F, M>(&mut self, masm: &mut M, size: OperandSize, mut emit: F)
```

Comment:
I don't think there's much benefit on adding this new function. The reason why the `i32` and `i64` variants for ints exist is that we can optimize cases in which an int32 or int64 are on top of the stack, and emit an instruction that takes an immediate rather than loading that value into a register. In the case of floats, at least in x64, no instruction supports a float immediate as operand, plus we are not doing any sort of type checking here either. My expectation is that we can refactor the existing binop implementations to share the general case (register operand) binop machinery across i32|i64 and floats. We could do so by extracting the `else` case of `i32_binop` and `i64_binop` into a general `binop` function, similar to `unop`, which is used by `f32.abs` for example.

Reply:
One thing that makes that a little complicated is the integer binops use an emit closure that takes a `RegImm` as an argument while the float ones take a `Reg` because they can't use an immediate. I can change the emit signature to take a `RegImm` and the change the visitor code to look like:

```rust
    fn visit_f32_add(&mut self) {
        self.context.binop(
            self.masm,
            OperandSize::S32,
            &mut |masm: &mut M, dst, src: RegImm, size| {
                masm.float_add(dst, dst, src.try_into().expect("register"), size);
            },
        );
    }
```

Or I could have the macroassembler methods accept `RegImm`s and have the x86 implementation perform the `try_into()` with an `expect`.

I could also look into using a generic for the `src` parameter type on the emit closure.

Thoughts on what direction would make the most sense to you?

Reply:
Another option that just occurred to me was to adapt the emit closures in the `i32` and `i64` methods. That approach looks the cleanest so I'll proceed with that one for now.

Reply:
I think that expecting a register at the visitor level, like in your code snippet is the approach that I'd prefer: that makes it clear what the invariant with float binary operations is.